File: packages/ts-sdk/src/parse.ts
import {
  validateWorkoutLog,
  validateWorkoutTemplate,
  validateProgram,
  validateLifterProfile,
} from './validate.js'
import type { WorkoutLog, WorkoutTemplate, Program, LifterProfile } from './types.js'

export class ParseError extends Error {
  constructor(
    message: string,
    public readonly errors: { path: string; message: string }[] = []
  ) {
    super(message)
    this.name = 'ParseError'
  }
}

// ============================================
// Workout Log Parsing
// ============================================

export function parseWorkoutLog(json: string): WorkoutLog {
  let data: unknown
  try {
    data = JSON.parse(json)
  } catch {
    throw new ParseError('Invalid JSON')
  }

  const result = validateWorkoutLog(data)
  if (!result.valid) {
    throw new ParseError('Schema validation failed', result.errors)
  }

  return data as WorkoutLog
}

// ============================================
// Workout Template Parsing
// ============================================

export function parseWorkoutTemplate(json: string): WorkoutTemplate {
  let data: unknown
  try {
    data = JSON.parse(json)
  } catch {
    throw new ParseError('Invalid JSON')
  }

  const result = validateWorkoutTemplate(data)
  if (!result.valid) {
    throw new ParseError('Schema validation failed', result.errors)
  }

  return data as WorkoutTemplate
}

// ============================================
// Program Parsing
// ============================================

export function parseProgram(json: string): Program {
  let data: unknown
  try {
    data = JSON.parse(json)
  } catch {
    throw new ParseError('Invalid JSON')
  }

  const result = validateProgram(data)
  if (!result.valid) {
    throw new ParseError('Schema validation failed', result.errors)
  }

  return data as Program
}

// ============================================
// Lifter Profile Parsing
// ============================================

export function parseLifterProfile(json: string): LifterProfile {
  let data: unknown
  try {
    data = JSON.parse(json)
  } catch {
    throw new ParseError('Invalid JSON')
  }

  const result = validateLifterProfile(data)
  if (!result.valid) {
    throw new ParseError('Schema validation failed', result.errors)
  }

  return data as LifterProfile
}

File: packages/kotlin-sdk/src/main/kotlin/com/openweight/Parse.kt
package com.openweight

import com.openweight.model.LifterProfile
import com.openweight.model.Program
import com.openweight.model.WorkoutLog
import com.openweight.model.WorkoutTemplate
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement

private val jsonParser = Json { ignoreUnknownKeys = true }

private inline fun <reified T> parse(
    jsonString: String,
    validate: (JsonElement) -> ValidationResult
): T {
    val element = try {
        Json.parseToJsonElement(jsonString)
    } catch (e: Exception) {
        throw ParseException("Invalid JSON: ${e.message}")
    }

    val result = validate(element)
    if (!result.valid) {
        throw ParseException("Schema validation failed", result.errors)
    }

    return jsonParser.decodeFromString<T>(jsonString)
}

/**
 * Parses a JSON string into a [WorkoutLog], validating against the schema.
 *
 * @param jsonString The JSON string to parse
 * @return The parsed [WorkoutLog]
 * @throws ParseException if the JSON is invalid or fails schema validation
 */
fun parseWorkoutLog(jsonString: String): WorkoutLog =
    parse(jsonString, ::validateWorkoutLog)

/**
 * Parses a JSON string into a [WorkoutTemplate], validating against the schema.
 *
 * @param jsonString The JSON string to parse
 * @return The parsed [WorkoutTemplate]
 * @throws ParseException if the JSON is invalid or fails schema validation
 */
fun parseWorkoutTemplate(jsonString: String): WorkoutTemplate =
    parse(jsonString, ::validateWorkoutTemplate)

/**
 * Parses a JSON string into a [Program], validating against the schema.
 *
 * @param jsonString The JSON string to parse
 * @return The parsed [Program]
 * @throws ParseException if the JSON is invalid or fails schema validation
 */
fun parseProgram(jsonString: String): Program =
    parse(jsonString, ::validateProgram)

/**
 * Parses a JSON string into a [LifterProfile], validating against the schema.
 *
 * @param jsonString The JSON string to parse
 * @return The parsed [LifterProfile]
 * @throws ParseException if the JSON is invalid or fails schema validation
 */
fun parseLifterProfile(jsonString: String): LifterProfile =
    parse(jsonString, ::validateLifterProfile)
